### PR TITLE
Add a 'kubeadm config print-default' command

### LINKED ISSUE #829: Add a command that prints the default API object YAML
For making it easier to write kubeadm YAML config. UX improvement and important now that we change the structure of the API

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Improves the UX around creating config files.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/829

**Special notes for your reviewer**:

**Release note**:

```release-note
kubeadm: A `kubeadm config print-default` command has now been added that you can use as a starting point when writing your own kubeadm configuration files
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @liztio

### CODE REVIEW

**Review by timothysc — CHANGES_REQUESTED** [review]
minor nits and a question.

File: cmd/kubeadm/app/cmd/config.go [hunk 1]
```diff
@@ -65,12 +72,59 @@ func NewCmdConfig(out io.Writer) *cobra.Command {
...
 
+	cmd.AddCommand(NewCmdConfigṔrintDefault(out))
 	cmd.AddCommand(NewCmdConfigUpload(out, &kubeConfigFile))
 	cmd.AddCommand(NewCmdConfigView(out, &kubeConfigFile))
 	cmd.AddCommand(NewCmdConfigImages(out))
 	return cmd
 }
 
+// NewCmdConfigṔrintDefault returns cobra.Command for "kubeadm config print-default" command
+func NewCmdConfigṔrintDefault(out io.Writer) *cobra.Command {
```
Comment (mattkelly) [inline cmd/kubeadm/app/cmd/config.go]: You used `Ṕ` here in `Print` but I think you wanted a `P`

File: cmd/kubeadm/app/cmd/join.go [hunk 2]
```diff
@@ -129,24 +126,24 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 }
 
 // NewValidJoin validates the command line that are passed to the cobra command
-func NewValidJoin(cfg *kubeadmapiv1alpha2.NodeConfiguration, args []string, skipPreFlight bool, cfgPath, featureGatesString string, ignorePreflightErrors []string) (*Join, error) {
+func NewValidJoin(cmd *cobra.Command, cfg *kubeadmapiv1alpha2.NodeConfiguration, args []string, skipPreFlight bool, cfgPath, featureGatesString string, ignorePreflightErrors []string) (*Join, error) {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: `join_test.go` needs a little change:
```
cmd := NewCmdJoin(&out)
join, err := NewValidJoin(cmd, cfg, tc.args, tc.skipPreFlight, tc.cfgPath, tc.featureGatesString, tc.ignorePreflightErrors)
//                         ^^ passing cmd as argument now
```

File: cmd/kubeadm/app/cmd/join.go [hunk 3]
```diff
@@ -205,47 +202,31 @@ type Join struct {
...
-	return &Join{cfg: cfg}, nil
-}
-
-// Validate validates mixed arguments passed to cobra.Command
-func (j *Join) Validate(cmd *cobra.Command) error {
-	if err := validation.ValidateMixedArguments(cmd.PersistentFlags()); err != nil {
-		return err
-	}
-	return validation.ValidateNodeConfiguration(j.cfg).ToAggregate()
+	return &Join{cfg: internalcfg}, nil
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: welp, there is also this issue in `join_test.go`:
```
invalid operation: join (variable of type *Join) has no field or method Validate
````
possibly this and the related tests has to be removed:
https://github.com/luxas/kubernetes/blob/469532c49917bbecaf9d9b29ff5da5a4aa57209c/cmd/kubeadm/app/cmd/join_test.go#L177-L181

**Review by chuckha — APPROVED** [review]
This looks good to me, just one question:

This is printing the internal version of the config, shouldn't it be printing the external version of the config instead? Users should be defining the api extension version of the config, not the internal one.

**Comment by luxas** [comment]
> This is printing the internal version of the config, shouldn't it be printing the external version of the config instead? Users should be defining the api extension version of the config, not the internal one.

@chuckha This is printing the external version of the config :sunglasses:. The line doing the magic is this one:

```go
kubeadmutil.MarshalToYamlForCodecs(internalcfg, kubeadmapiv1alpha2.SchemeGroupVersion, kubeadmscheme.Codecs)
```
Telling the API machinery framework to serialize it for the external v1alpha2 `GroupVersion`.
Not obvious I know, but kinda cool :wink:

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by neolit123** [comment]
@luxas 
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/63969/pull-kubernetes-verify/91406/

Chuck was getting this due to not generating the docs locally and including all files(?) i think it's the same PY stacktrace.
/cc @chuckha

**Comment by luxas** [comment]
Yup, that's right @neolit123, thanks! Ran `./hack/update-generated-docs.sh` now.
Readding lgtm/approved.

**Pull request merged**